File: medium/koko-eating-bananas.js
// https://www.youtube.com/watch?v=U2SozAs9RzA

/**
 * @param {number[]} piles
 * @param {number} h
 * @return {number}
 */
var minEatingSpeed = function (piles, h) {
  let start = 1;
  let end = Math.max(...piles);
  let result;

  while (start <= end) {
    const mid = (start + end) >>> 1;
    let hours = 0;

    for (const pile of piles) {
      hours += Math.ceil(pile / mid);
    }

    if (hours <= h) {
      result = mid;
      end = mid - 1;
    } else start = mid + 1;
  }

  return result;
};

File: medium/koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        s = 1
        e = max(piles)

        while s <= e:
            mid = (s + e) // 2
            
            hours = 0
            for i in range(len(piles)):
                hours += math.ceil(piles[i] / mid)

            if hours > h: s = mid + 1
            else: e = mid - 1

        return s